localstack/utils/bootstrap.py

```diff
@@ -764,36 +765,40 @@ def start_infra_in_docker_detached(console):
 def wait_container_is_ready(timeout: Optional[float] = None):
     """Blocks until the localstack main container is running and the ready marker has been printed."""
     container_name = config.MAIN_CONTAINER_NAME
+    started = time.time()
 
     def is_container_running():
         return DOCKER_CLIENT.is_container_running(container_name)
 
     if not poll_condition(is_container_running, timeout=timeout):
         return False
 
-    logfile = LocalstackContainer(container_name).logfile
+    stream = DOCKER_CLIENT.stream_container_logs(container_name)
 
-    ready = threading.Event()
-
-    def set_ready_if_marker_found(_line: str):
-        if _line == constants.READY_MARKER_OUTPUT:
-            ready.set()
-
-    # start a tail on the logfile
-    listener = FileListener(logfile, set_ready_if_marker_found)
-    listener.start()
+    # create a timer that will terminate the log stream after the remaining timeout
+    timer = None
+    if timeout:
+        waited = time.time() - started
+        remaining = timeout - waited
+        # check the rare case that the timeout has already been reached
+        if remaining <= 0:
+            stream.close()
+            return False
+        timer = threading.Timer(remaining, stream.close)
+        timer.start()
 
     try:
-        # but also check the existing log in case the container has been running longer
-        with open(logfile, "r") as fd:
-            for line in fd:
-                if constants.READY_MARKER_OUTPUT == line.strip():
-                    return True
-
-        # TODO: calculate remaining timeout
-        return ready.wait(timeout)
+        for line in stream:
+            line = line.decode("utf-8").strip()
+            if line == constants.READY_MARKER_OUTPUT:
+                return True
```

Comment:
The stream will never be cancelled in this case, right?
Is this an oversight or on purpose? (since the process is ended afterwards anyway maybe? still, this could leave an instance of Popen with `docker logs -f` open)

Reply:
yep that was a great catch! i put a `call_safe(stream.close)` in the finally block. `stream.close()` will call `popen.terminate()` which will do nothing if the process already returned, so safe to call on an already closed stream.